### DISCUSSION TITLE
Help with strictness and expressions

### DISCUSSION THREAD

**Discussion by rozaliev** [kcl-lang/kcl#1005] [Q&A]
Hi, I'm evaluating KCL for our configs and ran into some troubles from the start. KCL v0.7.4

1: Is there any way to express this without intermediate schema like `WithComponent`
```
schema Application:
    name: str
    components: WithComponents

schema WithComponent:
    name: str
    $type: str

schema WithComponents:
    [name: str]: WithComponent = {name: name}
```

2: Why does this compile and produce incomplete result given missing `$type`?
```
schema Application:
    name: str
    components: WithComponents

schema WithComponent:
    name: str
    $type: str

schema WithComponents:
    [name: str]: WithComponent = {name: name}

myapp = {
    components: {
        "webservice": {}
    }
}

myapps: [Application] = [myapp]
o = Application.instances()


> myapp:
>   components:
>     webservice: {}
> myapps:
> - components:
>     webservice:
>       name: webservice
> o:
> - components:
>     webservice:
>       name: webservice

```

3. Why does this compile?
```
schema Application:
    name: str
    components: WithComponents

schema WithComponent:
    name: str
    $type: str

schema WithComponents:
    [name: str]: WithComponent = {name: name}

myapp2: Application {
    name: "myapp2"
    components.ws: WithComponent {}
}

o = Application.instances()

> o:
> - name: myapp2
>   components:
>     ws:
>       name: ws
> myapp2:
>   name: myapp2
>   components:
>     ws:
>       name: ws

```

4. Why is it required that I add `WithComponent` annotation here, given that it is already specified in schema? 

```
myapp2: Application {
    name: "myapp2"
    components.ws: WithComponent {...}
}

# vs

myapp2: Application {
    name: "myapp2"
    components.ws: {...}
}
```

5. More weird stuff, that I can't wrap my head around. Why is s2 empty?

```
schema P:
    name: str

schema S[p: P]:
    myname = p.name

schema S2[p: P]:
    sub.myname = p.name


sp = {name = "sname"}
s = S(sp) {}
s2 = S2(sp) {}

> sp:
>   name: sname
> s:
>   myname: sname
> s2: {}

```

**Comment by Peefy** [comment]
1. We still do not support this feature. You can open an Issue to track it. ❤️
2. `myapp` is a dict type field that does not have a `type` field, so it does not output. When it is assigned to the `Application` type, a type conversion occurs.
3 & 4 & 5: There may be potential bugs in index signature at the KCL runtime, and I will fix it. Thank you for your feedback.

**Reply by rozaliev** [reply]
2: I'm a bit confused here, why wouldn't compiler reject this conversion given that `myapp` dict does not comply with required schema?

**Reply by Peefy** [reply]
Sorry, this was my negligence and it has been fixed in https://github.com/kcl-lang/kcl/pull/1009. The code to walk around can be written as follows:

```
schema Application:
    name: str
    components: WithComponents

schema WithComponent:
    name: str
    $type: str

schema WithComponents:
    [name: str]: WithComponent = {name: name}

myapp = Application {
    components: {
        "webservice": {}
    }
}

myapps: [Application] = [myapp]
o = Application.instances()
```

**Comment by Peefy** [comment]
>    3 & 4 & 5: There may be potential bugs in index signature at the KCL runtime, and I will fix it. Thank you for your feedback.

fixed by https://github.com/kcl-lang/kcl/pull/1007

**Reply by rozaliev** [reply]
Thanks! Gonna check it out once brew packages are published.

**Reply by Peefy** [reply]
We have released kcl v0.7.5 and v0.8.0-alpha.1. Thank you again for your feedback. ❤️
